# 4.9 Data Visualization with Python

# 1. Download the customer data set and add it to your “Original Data” folder.

Done

# 2. Create a new notebook in your “Scripts” folder for part 1 of this task.

Done

# 3. Import your analysis libraries, as well as your new customer data set as a dataframe.

In [8]:
#Import libraries
import pandas as pd
import numpy as np
import os
import matplotlib.pyplot as plt
import seaborn as sns
import scipy

In [9]:
#Import new customer data set as dataframe
new_cust_data = pd.read_csv('/Users/ariedlin/Desktop/Career Foundry Assignments/8-24 Instacart Blanket Analysis Real/2 Data/1 Original Data/customers.csv')

In [10]:
#Check head() to see what the new data looks like
new_cust_data.head(10)

user_id First Name    Surnam  Gender           STATE  Age date_joined  \
0    26711    Deborah  Esquivel  Female        Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female      New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male           Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female            Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female        Maryland   26    1/1/2017   
5   133128    Cynthia     Noble  Female        Kentucky   43    1/1/2017   
6   152052      Chris    Walton    Male         Montana   20    1/1/2017   
7   168851     Joseph   Hickman    Male  South Carolina   30    1/1/2017   
8    69965     Jeremy      Vang    Male           Texas   47    1/1/2017   
9    82820      Shawn     Chung    Male        Virginia   26    1/1/2017   

   n_dependants fam_status  income  
0             3    married  165665  
1             0     single   59285  
2             2    married   99568  
3             0     single   42049  
4             1    married   40374  
5             2    married   49643  
6             0     single   61746  
7             0     single   63712  
8             1    married  162432  
9             2    married   32072

In [11]:
#Check shape of new dataframe
new_cust_data.shape

(206209, 10)

# 4. Wrangle the data so that it follows consistent logic; for example, rename columns with illogical names and drop columns that don’t add anything to your analysis.

I do not see any uselss columns to drop in this data set.  All of the columns could prove useful for Instacart analysis.

In [14]:
#Changing name of column 'surnam' to 'last_name'
new_cust_data.rename(columns = {'Surnam' : 'Last Name'}, inplace = True)

In [15]:
#Check to make sure it worked
new_cust_data.head()

user_id First Name Last Name  Gender       STATE  Age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   n_dependants fam_status  income  
0             3    married  165665  
1             0     single   59285  
2             2    married   99568  
3             0     single   42049  
4             1    married   40374

In [16]:
#Changing column 'STATE' to 'State' to make column names more uniform
new_cust_data.rename(columns = {'STATE' : 'State'}, inplace = True)

In [17]:
#Check to make sure it worked
new_cust_data.head()

user_id First Name Last Name  Gender       State  Age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   n_dependants fam_status  income  
0             3    married  165665  
1             0     single   59285  
2             2    married   99568  
3             0     single   42049  
4             1    married   40374

# 5. Complete the fundamental data quality and consistency checks you’ve learned throughout this Achievement; for example, check for and address missing values and duplicates, and convert any mixed-type data.

# Mixed Type Data

In [20]:
##Check for mixed type data
for col in new_cust_data.columns:
    # Compare types in the column with the type of the first row
    weird = (new_cust_data[col].apply(type) != type(new_cust_data[col].iloc[0]))
    
    # If any row has a different type, print the column name
    if weird.any():
        print(col)

user_id
First Name
Age
n_dependants
income


In [21]:
# Print the column data type for each column
print(new_cust_data[['user_id', 'First Name', 'Age', 'n_dependants', 'income']].dtypes)

user_id          int64
First Name      object
Age              int64
n_dependants     int64
income           int64
dtype: object


All columns appear to have the correct type of data

# Missing Data

In [24]:
## Check for missing data
new_cust_data.isnull().sum()

user_id             0
First Name      11259
Last Name           0
Gender              0
State               0
Age                 0
date_joined         0
n_dependants        0
fam_status          0
income              0
dtype: int64

There are 11,259 values missing for first name.  I do not think First Name is very crucial to my data analysis, so I am going to leave those as null.

# Duplicate Data

In [27]:
#Create a subset for duplicate data
df_dups = new_cust_data[new_cust_data.duplicated()]

In [28]:
#Check shape of new subset to see how many duplicates were found
df_dups.shape

(0, 10)

In [29]:
#View the whole subset to ensure no dups
df_dups

Empty DataFrame
Columns: [user_id, First Name, Last Name, Gender, State, Age, date_joined, n_dependants, fam_status, income]
Index: []

There are no duplicates in this dataframe.

# 6. Combine your customer data with the rest of your prepared Instacart data. (Hint: Make sure the key columns are the same data type!)

In [32]:
#Import updated ords_prods_merge dataframe with new columns
ords_prods_merge = pd.read_pickle('/Users/ariedlin/Desktop/Career Foundry Assignments/8-24 Instacart Blanket Analysis Real/2 Data/2 Prepared Data/ords_prods_merge')

In [33]:
#Check new_customer_data head() to see all column names
new_cust_data.head()

user_id First Name Last Name  Gender       State  Age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   n_dependants fam_status  income  
0             3    married  165665  
1             0     single   59285  
2             2    married   99568  
3             0     single   42049  
4             1    married   40374

In [34]:
#Check ords_prods_merge head() to see all column names
ords_prods_merge.head()

Unnamed: 0_x  order_id  user_id eval_set  order_number  orders_dow  \
0             0   2539329        1    prior             1           2   
1             0   2539329        1    prior             1           2   
2             0   2539329        1    prior             1           2   
3             0   2539329        1    prior             1           2   
4             0   2539329        1    prior             1           2   

   order_hour_of_day  days_since_prior_order  product_id  add_to_cart_order  \
0                  8                     NaN         196                  1   
1                  8                     NaN       14084                  2   
2                  8                     NaN       12427                  3   
3                  8                     NaN       26088                  4   
4                  8                     NaN       26405                  5   

   ...    price_range_loc     busiest_day         busiest_days  \
0  ...  mid_range_product  Regularly busy  Regularly busy days   
1  ...  mid_range_product  Regularly busy  Regularly busy days   
2  ...  low_range_product  Regularly busy  Regularly busy days   
3  ...  low_range_product  Regularly busy  Regularly busy days   
4  ...  low_range_product  Regularly busy  Regularly busy days   

    busiest_hours  max_order  loyalty_flag  spending_avg spender_type  \
0  Average orders         10  New customer      6.367797  Low spender   
1  Average orders         10  New customer      6.367797  Low spender   
2  Average orders         10  New customer      6.367797  Low spender   
3  Average orders         10  New customer      6.367797  Low spender   
4  Average orders         10  New customer      6.367797  Low spender   

  customer_frequency customer_frequency_flag  
0               20.5   Non-frequent customer  
1               20.5   Non-frequent customer  
2               20.5   Non-frequent customer  
3               20.5   Non-frequent customer  
4               20.5   Non-frequent customer  

[5 rows x 28 columns]

In [35]:
#Check data type of user_id in new_cust_data 
print(new_cust_data['user_id'].dtype)

int64


In [36]:
#Check data type of user_id in ords_prods_merge
print(ords_prods_merge['user_id'].dtype)

int64


In [37]:
#Use df_merge function to combine ords_prods_merge and new_cust_data using user_id column to link them
ords_prods_cust_merge = ords_prods_merge.merge(new_cust_data, on='user_id', indicator='merge_status_new')

In [38]:
#Check to see if they merged correctly
ords_prods_cust_merge.head()

Unnamed: 0_x  order_id  user_id eval_set  order_number  orders_dow  \
0             0   2539329        1    prior             1           2   
1             0   2539329        1    prior             1           2   
2             0   2539329        1    prior             1           2   
3             0   2539329        1    prior             1           2   
4             0   2539329        1    prior             1           2   

   order_hour_of_day  days_since_prior_order  product_id  add_to_cart_order  \
0                  8                     NaN         196                  1   
1                  8                     NaN       14084                  2   
2                  8                     NaN       12427                  3   
3                  8                     NaN       26088                  4   
4                  8                     NaN       26405                  5   

   ...  First Name Last Name  Gender    State  Age  date_joined  n_dependants  \
0  ...       Linda    Nguyen  Female  Alabama   31    2/17/2019             3   
1  ...       Linda    Nguyen  Female  Alabama   31    2/17/2019             3   
2  ...       Linda    Nguyen  Female  Alabama   31    2/17/2019             3   
3  ...       Linda    Nguyen  Female  Alabama   31    2/17/2019             3   
4  ...       Linda    Nguyen  Female  Alabama   31    2/17/2019             3   

  fam_status income merge_status_new  
0    married  40423             both  
1    married  40423             both  
2    married  40423             both  
3    married  40423             both  
4    married  40423             both  

[5 rows x 38 columns]

In [39]:
#Check shape to ensure same amount of rows as ords_prods_merge dataframe
ords_prods_cust_merge.shape

(32404859, 38)

In [40]:
ords_prods_merge.shape

(32404859, 28)

# 7. Ensure your notebook contains logical titles, section headings, and descriptive code comments.


In [42]:
#Check all column names in new dataframe
print(ords_prods_cust_merge.columns)

Index(['Unnamed: 0_x', 'order_id', 'user_id', 'eval_set', 'order_number',
       'orders_dow', 'order_hour_of_day', 'days_since_prior_order',
       'product_id', 'add_to_cart_order', 'reordered', '_merge',
       'Unnamed: 0_y', 'product_name', 'aisle_id', 'department_id', 'prices',
       'merge_status', 'price_range_loc', 'busiest_day', 'busiest_days',
       'busiest_hours', 'max_order', 'loyalty_flag', 'spending_avg',
       'spender_type', 'customer_frequency', 'customer_frequency_flag',
       'First Name', 'Last Name', 'Gender', 'State', 'Age', 'date_joined',
       'n_dependants', 'fam_status', 'income', 'merge_status_new'],
      dtype='object')


In [43]:
#Remove unnecessary columns from the dataframe
ords_prods_cust_merge_ = ords_prods_cust_merge.drop(columns=['Unnamed: 0_x', 'eval_set', '_merge', 'Unnamed: 0_y', 'merge_status'])

In [44]:
#Check all column names in final dataframe
print(ords_prods_cust_merge_.columns)

Index(['order_id', 'user_id', 'order_number', 'orders_dow',
       'order_hour_of_day', 'days_since_prior_order', 'product_id',
       'add_to_cart_order', 'reordered', 'product_name', 'aisle_id',
       'department_id', 'prices', 'price_range_loc', 'busiest_day',
       'busiest_days', 'busiest_hours', 'max_order', 'loyalty_flag',
       'spending_avg', 'spender_type', 'customer_frequency',
       'customer_frequency_flag', 'First Name', 'Last Name', 'Gender', 'State',
       'Age', 'date_joined', 'n_dependants', 'fam_status', 'income',
       'merge_status_new'],
      dtype='object')


In [45]:
ords_prods_cust_merge_.head()

order_id  user_id  order_number  orders_dow  order_hour_of_day  \
0   2539329        1             1           2                  8   
1   2539329        1             1           2                  8   
2   2539329        1             1           2                  8   
3   2539329        1             1           2                  8   
4   2539329        1             1           2                  8   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                     NaN       14084                  2          0   
2                     NaN       12427                  3          0   
3                     NaN       26088                  4          0   
4                     NaN       26405                  5          0   

                              product_name  ...  First Name  Last Name  \
0                                     Soda  ...       Linda     Nguyen   
1  Organic Unsweetened Vanilla Almond Milk  ...       Linda     Nguyen   
2                      Original Beef Jerky  ...       Linda     Nguyen   
3               Aged White Cheddar Popcorn  ...       Linda     Nguyen   
4         XL Pick-A-Size Paper Towel Rolls  ...       Linda     Nguyen   

   Gender    State Age date_joined n_dependants  fam_status income  \
0  Female  Alabama  31   2/17/2019            3     married  40423   
1  Female  Alabama  31   2/17/2019            3     married  40423   
2  Female  Alabama  31   2/17/2019            3     married  40423   
3  Female  Alabama  31   2/17/2019            3     married  40423   
4  Female  Alabama  31   2/17/2019            3     married  40423   

   merge_status_new  
0              both  
1              both  
2              both  
3              both  
4              both  

[5 rows x 33 columns]

# 8. Export this new dataframe as a pickle file so you can continue to use it in the second part of this task.

In [47]:
#Export ords_prods_merge_ as a pickle
ords_prods_cust_merge_.to_pickle('/Users/ariedlin/Desktop/Career Foundry Assignments/8-24 Instacart Blanket Analysis Real/2 Data/2 Prepared Data/ords_prods_cust_merge_')